README.md
# AtmCircLib: Python library of the Atmospheric Circulation Group

## Start developing

Once you created or cloned this repository, make sure the installation is running properly. Install the package dependencies with the provided script `setup_env.sh`.
Check available options with
```bash
tools/setup_env.sh -h
```
We distinguish development installations which are editable and have additional dependencies on formatters and linters from productive installations which are non-editable
and have no additional dependencies. Moreover we distinguish pinned installations based on exported (reproducible) environments and free installations where the installation
is based on top-level dependencies listed in `requirements.yml`. If you start developing, you might want to do an unpinned
development installation and export the environments right away.

```bash
tools/setup_env.sh -u -d -e -n <package_env_name>
```
*Hint*: If you are the package administrator, it is a good idea to understand what this script does, you can do everything manually with `conda` instructions.

*Hint*: Use the flag `-m` to speed up the installation using mamba. Of course you will have to install mamba first (we recommend to install mamba into your base
environment `conda install -c conda-forge mamba`. If you install mamba in another (maybe dedicated) environment, environments installed with mamba will be located
in `<miniconda_root_dir>/envs/mamba/envs`, which is not very practical.

The package itself is installed with `pip`:

```bash
conda activate <package_env_name>
pip install --editable .
```

*Warning:* Make sure you use the right pip, i.e. the one from the installed conda environment (`which pip` should point to something like `path/to/miniconda/envs/<package_env_name>/bin/pip`).

Once your package is installed, run the tests by typing:

```
conda activate <package_env_name>
pytest
```

If the tests pass, you are good to go. If not, contact the package administrator . Make sure to update your requirement files and export your environments after installation
every time you add new imports while developing. Check the next section to find some guidance on the development process if you are new to Python and/or APN.

### Roadmap to your first contribution

Generally, the source code of your library is located in `src/<library_name>`. The blueprint will generate some example code in `mutable_number.py`, `utils.py` and `cli.py`. `cli.py` thereby serves as an entry
point for functionalities you want to execute from the command line, it is based on the Click library. If you do not need interactions with the command line, you should remove `cli.py`. Moreover, of course there exist other options for command line interfaces,
a good overview may be found here (https://realpython.com/comparing-python-command-line-parsing-libraries-argparse-docopt-click/), we recommend however to use click. The provided example
code should provide some guidance on how the individual source code files interact within the library. In addition to the example code in `src/<library_name>`, there are examples for
unit tests in `tests/<library_name>/`, which can be triggered with `pytest` from the command line. Once you implemented a feature (and of course you also
implemented a meaningful test ;-)), you are likely willing to commit it. First, go to the root directory of your package and run pytest.

```bash
conda activate <package_env_name>
cd <package-root-dir>
pytest
```

Note that neither pytest, nor pre-commit, nor any of the linters invoked by the pre-commit hooks will be available in the production environment, so make sure you have a development environment
installed and activated. If you use the blueprint as is, pre-commit will not be triggered locally but only if you push to the main branch
(or push to a PR to the main branch). If you consider it useful, you can set up pre-commit to run locally before every commit by initializing it once. In the root directory of
your package, type:

```bash
pre-commit install
```

If you run `pre-commit` without installing it before (line above), it will fail and the only way to recover it, is to do a forced reinstallation (`conda install --force-reinstall pre-commit`).
You can also just run pre-commit selectively, whenever you want by typing (`pre-commit run --all-files`). Note that mypy and pylint take a bit of time, so it is really
up to you, if you want to use pre-commit locally or not. In any case, after running pytest, you can commit and the linters will run at the latest on the GitHub actions server,
when you push your changes to the main branch. Note that pytest is currently not invoked by pre-commit, so it will not run automatically. Automated testing should be implemented
in a Jenkins pipeline (template for a plan available in `jenkins/`. See the next section for more details.

## Development tools

As this package was created with the APN Python blueprint, it comes with a stack of development tools, which are described in more detail on
(https://meteoswiss-apn.github.io/mch-python-blueprint/). Here, we give a brief overview on what is implemented.

### Testing and coding standards

Testing your code and compliance with the most important Python standards is a requirement for Python software written in APN. To make the life of package
administrators easier, the most important checks are run automatically on GitHub actions. If your code goes into production, it must additionally be tested on CSCS
machines, which is only possible with a Jenkins pipeline (GitHub actions is running on a GitHub server).

### Pre-commit on GitHub actions

`.github/workflows/pre-commit.yml` contains a hook that will trigger the creation of your environment (unpinned) on the GitHub actions server and
then run pytest as well as various formatters and linters through pre-commit. This hook is only triggered upon pushes to the main branch (in general: don't do that)
and in pull requests to the main branch.

### Jenkins

Two jenkins plans are available in the `jenkins/` folder. On the one hand `jenkins/Jenkinsfile` controls the nightly (weekly, monthly, ...) builds, on the other hand
`jenkins/JenkinsJobPR` controls the pipeline invoked with the command `launch jenkins` in pull requests on GitHub. Your jenkins pipeline will not be set up
automatically. If you need to run your tests on CSCS machines, contact DevOps to help you with the setup of the pipelines. Otherwise, you can ignore the jenkinsfiles
and exclusively run your tests and checks on GitHub actions.

## Features

- TODO

## Credits

This package was created with [`copier`](https://github.com/copier-org/copier) and the [`MeteoSwiss-APN/mch-python-blueprint`](https://meteoswiss-apn.github.io/mch-python-blueprint/) project template.


HISTORY.md
# AtmCircLib: History

## v0.9.2 (2023-01-05)

- Lower min. typer version further to 0.4 to accommodate rstcheck (requires <0.5, not <=0.5)

## v0.9.1 (2023-01-04)

- Fix editable installs for extension modules (broken in template when stripping down setup.py)
- Switch copier template to new `atmcirc/python-blueprint-f2py` based on `mch-python-blueprint`
  - Adds docs and setups for github actions and jenkins
- Reduce min. typer version to 0.5 (otherwise incompatible with rstcheck)
- Merge `dev-requirements.yml` into `requirements.yml` and eliminate `requirements/dev*.yml`

## v0.9.0 (2022-12-19)

- In `atmcirclib.regular_grid.RegularGridPlot`, implement optional rasterization of topography
- Major copier update: Move toward the MCH Blueprint (not quite finished)

## v0.8.0 (2022-10-12)

- Add module `atmcirclib.plot_utils` with functions to manipulate colormaps (moved from `cosmo_natl.crclim_domains`).
- Add module `atmcirclib.regular_grid` with classes `RegularGrid`, `RegularGridPlotter` and `RegularGridPlot` (originally moved from `cosmo_natl.crclim_domains`)
  - Purpose: Create plot with multiple grid outlines on top of the model topography

## v0.7.1 (2022-10-05)

- Run copier updates
  - Replaces `{,dev}-requirements.in` by `~.yml`
  - Moves requirements and environments files into `requirements/`
- Update `gt4py` version to `v2022-09-29`
  - Fixes installation problem with `xxhash` (called `python-xxhash` in `conda-forge`)

## v0.7.0 (2022-09-14)

- Add module `atmcirclib.typer` with decorator `typer_wrap_pdb` (analogue to `atmcirclib.click.click_wrap_pdb`) and some additional conveniences
- Add subpackage `atmcirclib.fortran` with Fortran namelist parser with `parse_namelist` and `parse_namelist_file`
- Add module `atmcirclib.icon` with various utilities
  - Find path to full namelist by parsing master namelist file in ICON run directory with `find_namelist_path`
  - Insert ICON namelist params in format string templates with `format_icon_params`
  - Convert time steps from ICON format to regular datetime objects with `convert_icon_time_step` and `convert_icon_time`
  - Disaggregate total precip with `disaggr_tot_precip` and derive precipitation intensity units with `deriv_precip_intensity_units`
- Add module `atmcirclib.math` with function `step_ceil` with tests
- Fix `tools/create_envs.sh` (`${UPDATE}` implemented wrong)
  - Merged back into copier template `meta-python-project`
- Run copier update (template `python-project`)
  - `tools/create_envs.sh`: Fixed `${UPDATE}` bug; don't update by default
  - `tools/install_editable.sh`: Refactored; build extension modules by default; don't install linked by default

## v0.6.0 (2022-08-29)

- Add callback function `click_set_pdb_raise` to `atmcirclib.click`
- Add decorator function `click_add_option_pdb` to `atmcirclib.click` to add `--pdb` option with one line
- Add decorator function `click_wrap_pdb` to wrap click command function in `wrap_pdb` with option to call `click_add_option_pdb` internally, making the addition of `--pdb` a (potential) one-liner
- Update conda environments
- Update to copier template

## v0.5.0 (2022-06-28)

- Hook project back up to copier template python-project-f2py and run copier update
- Write simulation scanner to find simulation run directories
- Improve simulations code while working on simulations library in `cosmo-nawdex` repo
- Add module `plot.rgb` from `cosmo-nawdex` and adapt it

## v0.4.0 (2022-06-08)

- Improve functions used to drop into pdb in case of exception in `click` module, incl. docstrings
- Add module `simulations` (from `cosmo_nawdex` repo)
- Add new requirement `pandas` and update envs
- Remove conda build file archive; only retain latest as `recipe/meta.yaml`

## v0.3.0 (2022-05-31)

- Move `atmcirclib.tools.call_graph` and `atmcirclib.tools.create_recipe` to `atmcirctools v0.2.0` as commands `act call-graph` and `act create-recipe`
- Remove module `atmcirclib.tools`
- Merge `tools/{,run_}update_min_versions.sh` as `tools/update_min_reqs_versions.sh` and implement handling of comments

## v0.2.4 (2022-05-23)

- Add additional functions to module `click` (from `pyflexplot` via `cosmo_nawdex`)
  - Wrappers to drop into `ipdb` if exception/callback raises an exception
- Fix issue with `f2py` include path in `CMakeLists.txt`
- Write pair of bash scripts (`tools/{,run_}update_min_versions.sh`) to set min. versions in `requirements.in` to minor version of currently installed version
- Use `mamba` in `tox` to create envs
- Write script (`tools/call_graph.py`) to plot the call graph of a python module/package (based on `pycg` and `pygraphviz`)

## v0.2.3 (2022-04-25)

- Add module `click`

## v0.2.2 (2022-04-14)

- Add module `intp` (from repo `dpv_th`) with tests
- Add explorative `gt4py` tests (from repo `dpv_th`)
- Improve scripts to create recipe (now moved into atmcirclib) and environments

## v0.2.1 (2022-04-14)

- Switch build from (pure) `setuptools` to `scikit-build`
- Add Fortran extension modules as `atmcirclib.deriv.ext` (`deriv-f`, `deriv.f90`)

## v0.2.0 (2022-04-05)

- Implement `CriteriaFormatter` with string formatting of `*Criterion` and `Criteria` instances for titles ("human") and file names ("file"), with tests
- Rename `TrajDatasetMetadata` to `TrajTimeHandler`, extend it with some functionality required in `cosmo_nawdex.bin.plot_traj_density` and add tests for some crucial methods
- Restructure code and test packages
- Implement reading LAGRANTO output (NetCDF) and start files (four columns, no header)
- Implement creation of cosmo grid dataset from traj dataset
- Add model (`TrajModel`) and integration direction (`TrajDirection`) to `TrajDataset`
- Usual cleanup and refactoring along the way

### TODOs

- [ ] Add tests for untested `TrajTimeHandler` methods

## v0.1.0 (2022-03-29)

- Break up monolithic `TrajsDataset` (now `TrajsDataset`) class
- Move handling of COSMO grid file to new class `COSMOGridDataset`

src/atmcirclib/cosmo.py
"""COSMO output files."""
from __future__ import annotations

# Standard library
from typing import Any
from typing import TYPE_CHECKING

# Third-party
import cartopy.crs as ccrs  # type: ignore [import]
import numpy as np
import numpy.typing as npt
import xarray as xr

# First-party
from atmcirclib.geo import BoundingBox
from atmcirclib.geo import unrotate_coords
from atmcirclib.typing import PathLike_T

if TYPE_CHECKING:
    # First-party
    from atmcirclib.traj import TrajDataset


class COSMOGridDataset:
    """File with grid information of COSMO simulation."""

    def __init__(self, ds: xr.Dataset, z_unit: str = "m") -> None:
        """Create new instance."""
        self.ds: xr.Dataset = ds
        self.z_unit: str = z_unit

    def get_bbox_xy(self) -> BoundingBox:
        """Get (lon, lat) bounding box."""
        return BoundingBox.from_coords(self.ds.rlon.data, self.ds.rlat.data)

    def get_bbox_xz(self) -> BoundingBox:
        """Get (lon, z) bounding box."""
        bbox_xy = self.get_bbox_xy()
        zmin = 0.0
        zmax = float(self.ds.height_toa.data)
        return BoundingBox(
            llx=bbox_xy.llx,
            urx=bbox_xy.urx,
            lly=self._scale_z(zmin),
            ury=self._scale_z(zmax),
        )

    def get_bbox_yz(self) -> BoundingBox:
        """Get (lat, z) bounding box."""
        bbox_xy = self.get_bbox_xy()
        bbox_xz = self.get_bbox_xz()
        return BoundingBox(
            llx=bbox_xy.lly,
            urx=bbox_xy.ury,
            lly=bbox_xz.lly,
            ury=bbox_xz.ury,
        )

    def get_proj(self) -> ccrs.Projection:
        """Get domain projection."""
        # pylint: disable=E0110  # abstract-class-instantiated (RotatedPole)
        return ccrs.RotatedPole(
            pole_latitude=self.ds.rotated_pole.grid_north_pole_latitude,
            pole_longitude=self.ds.rotated_pole.grid_north_pole_longitude,
        )

    def _scale_z(self, val: float) -> float:
        """Scale height value to desired unit ``self.z_unit``."""
        unit = self.ds.height_toa.attrs["units"]
        if self.z_unit == unit:
            fact = 1.0
        elif self.z_unit == "km" and unit == "m":
            fact = 1 / 1000.0
        elif self.z_unit == "m" and unit == "km":
            fact = 1000.0
        else:
            raise Exception(
                f"unknown target z unit '{self.z_unit}' or input z unit '{unit}'"
            )
        return fact * val

    @classmethod
    def from_file(cls, path: PathLike_T, **kwargs: Any) -> COSMOGridDataset:
        """Create an instance from a NetCDF file."""
        # mypy thinks return type is Any (mypy v0.941, numpy v1.22.3)
        try:
            ds: xr.Dataset = xr.open_dataset(path, engine="h5netcdf")
        except (IOError, TypeError) as e:
            raise Exception(f"error reading cosmo grid file '{path}'") from e
        return cls(ds, **kwargs)

    @classmethod
    def from_trajs(
        cls,
        trajs: TrajDataset,
        *,
        pole_lon: float = 180.0,
        pole_lat: float = 90.0,
        **kwargs: Any,
    ) -> COSMOGridDataset:
        """Create an instance from a trajectories dataset."""
        rlon_min = np.nanmin(trajs.get_data("longitude"))
        rlon_max = np.nanmax(trajs.get_data("longitude"))
        rlat_min = np.nanmin(trajs.get_data("latitude"))
        rlat_max = np.nanmax(trajs.get_data("latitude"))
        z_max = np.nanmax(trajs.get_data("z"))
        ds = create_cosmo_grid_dataset_ds(
            rlon=np.linspace(rlon_min, rlon_max, 11).astype(np.float32),
            rlat=np.linspace(rlat_min, rlat_max, 11).astype(np.float32),
            pole_rlon=float(pole_lon),
            pole_rlat=float(pole_lat),
            height_toa_m=float(z_max),
        )
        return cls(ds, **kwargs)


def create_cosmo_grid_dataset_ds(
    rlon: npt.NDArray[np.float_],
    rlat: npt.NDArray[np.float_],
    *,
    pole_rlon: float = 180.0,
    pole_rlat: float = 90.0,
    height_toa_m: float = 20_000,
) -> xr.Dataset:
    """Create a dataset representing a COSMO file with grids."""

    def create_coord_time() -> xr.DataArray:
        """Create coordinate variable ``time``."""
        name = "time"
        # mypy 0.941 doesn't recognize return type of np.array (numpy 1.22.3)
        data: npt.NDArray[np.generic] = np.array(
            ["2016-09-20T00:00:00.000000000"], dtype="datetime64[ns]"
        )
        return xr.DataArray(
            name=name,
            data=data,
            dims=(name,),
            coords={name: data},
            attrs={
                "standard_name": "time",
                "long_name": "time",
                "bounds": "time_bnds",
            },
        )

    def create_coord_rlon() -> xr.DataArray:
        """Create coordinate variable ``rlon``."""
        name = "rlon"
        return xr.DataArray(
            name=name,
            data=rlon,
            dims=(name,),
            coords={name: rlon},
            attrs={
                "standard_name": "grid_longitude",
                "long_name": "rotated_longitude",
                "units": "degrees",
            },
        )

    def create_coord_rlat() -> xr.DataArray:
        """Create coordinate variable ``rlat``."""
        name = "rlat"
        return xr.DataArray(
            name=name,
            data=rlat,
            dims=(name,),
            coords={name: rlat},
            attrs={
                "standard_name": "grid_latitude",
                "long_name": "rotated_latitude",
                "units": "degrees",
            },
        )

    def create_coord_lon_lat() -> tuple[xr.DataArray, xr.DataArray]:
        """Create regular variables ``lon`` and ``lat``."""
        rlon = create_coord_rlon()
        rlat = create_coord_rlat()
        rotated_pole = create_rotated_pole()
        lon_arr, lat_arr = unrotate_coords(
            rlon=rlon.data,
            rlat=rlat.data,
            pole_rlon=rotated_pole.attrs["grid_north_pole_longitude"],
            pole_rlat=rotated_pole.attrs["grid_north_pole_latitude"],
            transpose=True,
        )
        dims = ("rlat", "rlon")
        coords = {
            "rlon": rlon.data,
            "rlat": rlat.data,
            "lon": (dims, lon_arr),
            "lat": (dims, lat_arr),
        }
        # mypy 0.941 doesn't recognize return type of xr.DataArray (2022.3.0)
        lon: xr.DataArray = xr.DataArray(
            name="lon",
            data=lon_arr,
            coords=coords,
            dims=dims,
            attrs={
                "standard_name": "longitude",
                "long_name": "longitude",
                "units": "degrees_east",
            },
        )
        # mypy 0.941 doesn't recognize return type of xr.DataArray (2022.3.0)
        lat: xr.DataArray = xr.DataArray(
            name="lat",
            data=lat_arr,
            coords=coords,
            dims=dims,
            attrs={
                "standard_name": "latitude",
                "long_name": "latitude",
                "units": "degrees_north",
            },
        )
        return (lon, lat)

    def create_rotated_pole() -> xr.DataArray:
        """Create regular variable ``rotated_pole``."""
        return xr.DataArray(
            name="rotated_pole",
            data=np.array([], dtype="S1"),
            attrs={
                "long_name": "coordinates of the rotated North Pole",
                "grid_mapping_name": "rotated_latitude_latitude",
                "grid_north_pole_latitude": pole_rlat,
                "grid_north_pole_longitude": pole_rlon,
            },
        )

    def create_time_bnds() -> xr.DataArray:
        """Create regular variable ``time_bnds``."""
        time = create_coord_time()
        return xr.DataArray(
            name="time_bnds",
            data=np.array([time.data[0], time.data[-1]]),
            dims=("bnds",),
            attrs={"long_name": "time bounds"},
        )

    def create_height_toa() -> xr.DataArray:
        """Create regular variable ``height_toa``."""
        return xr.DataArray(
            name="height_toa",
            data=np.array(height_toa_m, np.float32),
            attrs={
                "standard_name": "height",
                "long_name": "height of top of model",
                "units": "m",
                "axis": "Z",
                "positive": "up",
            },
        )

    coords: dict[str, xr.DataArray] = {}
    data_vars: dict[str, xr.DataArray] = {}

    coords["time"] = create_coord_time()
    coords["rlon"] = create_coord_rlon()
    coords["rlat"] = create_coord_rlat()
    coords["lon"], coords["lat"] = create_coord_lon_lat()

    data_vars["rotated_pole"] = create_rotated_pole()
    data_vars["time_bnds"] = create_time_bnds()
    data_vars["height_toa"] = create_height_toa()

    attrs = {
        "Conventions": "CF-1.4",
        "conventionsURL": "http://www.cfconventions.org/",
        "creation_date": "2021-12-05 19:03:25",
    }

    return xr.Dataset(
        coords=coords,
        data_vars=data_vars,
        attrs=attrs,
    )
